File: backend/src/main/java/com/smartlock/service/CalendarEngine.java
package com.smartlock.service;

import com.smartlock.domain.BookingHold;
import com.smartlock.domain.CalendarInterval;
import com.smartlock.domain.HostVerification;
import com.smartlock.domain.Property;
import com.smartlock.domain.PropertyBlockedDate;
import com.smartlock.domain.enums.PropertyStatus;
import com.smartlock.exception.AppException;
import com.smartlock.repository.BookingHoldRepository;
import com.smartlock.repository.CalendarIntervalRepository;
import com.smartlock.repository.HostVerificationRepository;
import com.smartlock.repository.PropertyBlockedDateRepository;
import com.smartlock.repository.PropertyRepository;
import jakarta.persistence.EntityManager;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Duration;
import java.time.Instant;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

/**
 * Core calendar state engine for Propvian.
 *
 * Responsibilities:
 *  - Availability checks (range-based, fast)
 *  - Booking holds (temporary RESERVED state with TTL)
 *  - State transitions (hold → booked, booked → cancelled, etc.)
 *  - Buffer interval management (auto-block before/after bookings)
 *  - Concurrency safety via PostgreSQL advisory locks + GIST exclusion constraint
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class CalendarEngine {

    private static final Duration HOLD_TTL = Duration.ofMinutes(15);

    private final CalendarIntervalRepository intervalRepo;
    private final BookingHoldRepository holdRepo;
    private final PropertyRepository propertyRepo;
    private final PropertyRuleResolver ruleResolver;
    private final HostVerificationRepository hostVerificationRepo;
    private final PropertyBlockedDateRepository blockedDateRepo;
    private final EntityManager em;

    // ── Public record types ───────────────────────────────────────────────────

    public record AvailabilityResult(boolean available, String reason) {
        static AvailabilityResult ok()               { return new AvailabilityResult(true,  null); }
        static AvailabilityResult blocked(String r)  { return new AvailabilityResult(false, r);    }
    }

    public record HoldResult(UUID holdId, UUID intervalId, Instant expiresAt) {}

    // ── Availability check ────────────────────────────────────────────────────

    @Transactional(readOnly = true)
    public AvailabilityResult checkAvailability(UUID propertyId, LocalDate checkIn, LocalDate checkOut) {
        validateDates(checkIn, checkOut);

        Property property = requireActiveProperty(propertyId);
        validateStayRules(property, checkIn, checkOut);

        List<CalendarInterval> conflicts = intervalRepo.findBlockingOverlap(propertyId, checkIn, checkOut);
        if (!conflicts.isEmpty()) {
            String states = conflicts.stream().map(CalendarInterval::getState).distinct()
                    .reduce((a, b) -> a + "," + b).orElse("UNKNOWN");
            return AvailabilityResult.blocked("Dates conflict with existing " + states + " intervals");
        }
        if (!blockedDateRepo.findOverlapping(propertyId, checkIn, checkOut).isEmpty()) {
            return AvailabilityResult.blocked("Dates are blocked by the host");
        }
        return AvailabilityResult.ok();
    }

    // ── Calendar view (website builder + admin calendar UI) ──────────────────

    @Transactional(readOnly = true)
    public List<CalendarInterval> getCalendar(UUID propertyId, LocalDate windowStart, LocalDate windowEnd) {
        return intervalRepo.findInWindow(propertyId, windowStart, windowEnd);
    }

    // ── Hold creation (guest begins checkout) ────────────────────────────────

    /**
     * Acquires a PostgreSQL advisory lock on the property, validates availability,
     * then atomically inserts a RESERVED interval + booking_hold record.
     * The advisory lock is transaction-scoped and released automatically on commit.
     */
    @Transactional
    public HoldResult createHold(UUID propertyId, LocalDate checkIn, LocalDate checkOut,
                                  int guests, String guestName, String guestEmail, String sessionId) {
        validateDates(checkIn, checkOut);
        Property property = requireActiveProperty(propertyId);
        validateStayRules(property, checkIn, checkOut);

        acquireAdvisoryLock(propertyId);

        // Re-check inside lock scope (TOCTOU prevention)
        List<CalendarInterval> conflicts = intervalRepo.findBlockingOverlap(propertyId, checkIn, checkOut);
        if (!conflicts.isEmpty()) {
            throw new AppException("Property is not available for the selected dates", HttpStatus.CONFLICT);
        }

        Instant expiresAt = Instant.now().plus(HOLD_TTL);

        CalendarInterval interval = CalendarInterval.builder()
                .propertyId(propertyId)
                .startDate(checkIn)
                .endDate(checkOut)
                .state("RESERVED")
                .expiresAt(expiresAt)
                .note("Guest hold: " + (guestEmail != null ? guestEmail : sessionId))
                .build();
        interval = intervalRepo.save(interval);

        BookingHold hold = BookingHold.builder()
                .propertyId(propertyId)
                .intervalId(interval.getId())
                .guestName(guestName)
                .guestEmail(guestEmail)
                .guestSessionId(sessionId)
                .checkinDate(checkIn)
                .checkoutDate(checkOut)
                .numberOfGuests(guests)
                .expiresAt(expiresAt)
                .status("ACTIVE")
                .build();
        hold = holdRepo.save(hold);

        // Link interval back to hold
        interval.setHoldId(hold.getId());
        intervalRepo.save(interval);

        log.info("Hold created: property={} hold={} [{} → {}] expires={}",
                propertyId, hold.getId(), checkIn, checkOut, expiresAt);

        return new HoldResult(hold.getId(), interval.getId(), expiresAt);
    }

    // ── Confirm hold → booked (payment succeeded) ────────────────────────────

    @Transactional
    public CalendarInterval confirmHold(UUID holdId, UUID bookingId) {
        BookingHold hold = holdRepo.findById(holdId)
                .orElseThrow(() -> new AppException("Hold not found", HttpStatus.NOT_FOUND));

        if (!"ACTIVE".equals(hold.getStatus())) {
            throw new AppException("Hold is no longer active (status: " + hold.getStatus() + ")",
                    HttpStatus.CONFLICT);
        }
        if (hold.getExpiresAt().isBefore(Instant.now())) {
            throw new AppException("Hold has expired", HttpStatus.CONFLICT);
        }

        CalendarInterval interval = intervalRepo.findById(hold.getIntervalId())
                .orElseThrow(() -> new AppException("Interval not found for hold", HttpStatus.INTERNAL_SERVER_ERROR));

        interval.setState("BOOKED");
        interval.setBookingId(bookingId);
        interval.setExpiresAt(null);
        interval = intervalRepo.save(interval);

        hold.setStatus("CONVERTED");
        holdRepo.save(hold);

        insertBuffers(interval);

        log.info("Hold confirmed: hold={} booking={} [{} → {}]",
                holdId, bookingId, interval.getStartDate(), interval.getEndDate());

        return interval;
    }

    // ── Release hold (payment failed / guest abandoned) ──────────────────────

    @Transactional
    public void releaseHold(UUID holdId) {
        BookingHold hold = holdRepo.findById(holdId)
                .orElseThrow(() -> new AppException("Hold not found", HttpStatus.NOT_FOUND));

        if ("CONVERTED".equals(hold.getStatus())) {
            throw new AppException("Cannot release a converted hold", HttpStatus.BAD_REQUEST);
        }

        if (hold.getIntervalId() != null) {
            intervalRepo.findById(hold.getIntervalId()).ifPresent(intervalRepo::delete);
        }

        hold.setStatus("RELEASED");
        holdRepo.save(hold);

        log.info("Hold released: hold={} property={}", holdId, hold.getPropertyId());
    }

    // ── Block / unblock dates (owner action) ─────────────────────────────────

    @Transactional
    public CalendarInterval blockDates(UUID propertyId, LocalDate startDate, LocalDate endDate,
                                        UUID actorId, String reason) {
        validateDates(startDate, endDate);
        acquireAdvisoryLock(propertyId);

        List<CalendarInterval> conflicts = intervalRepo.findBlockingOverlap(propertyId, startDate, endDate);
        if (!conflicts.isEmpty()) {
            throw new AppException("Date range conflicts with existing intervals", HttpStatus.CONFLICT);
        }

        CalendarInterval block = CalendarInterval.builder()
                .propertyId(propertyId)
                .startDate(startDate)
                .endDate(endDate)
                .state("BLOCKED")
                .blockedBy(actorId)
                .note(reason)
                .build();

        return intervalRepo.save(block);
    }

    @Transactional
    public void unblockDates(UUID intervalId, UUID actorId) {
        CalendarInterval interval = intervalRepo.findById(intervalId)
                .orElseThrow(() -> new AppException("Interval not found", HttpStatus.NOT_FOUND));

        if (!"BLOCKED".equals(interval.getState())) {
            throw new AppException("Can only unblock BLOCKED intervals", HttpStatus.BAD_REQUEST);
        }

        intervalRepo.delete(interval);
        log.info("Dates unblocked: interval={} by={}", intervalId, actorId);
    }

    // ── Cancel booking (restore dates) ────────────────────────────────────────

    @Transactional
    public void cancelBookingDates(UUID bookingId) {
        List<CalendarInterval> intervals = intervalRepo.findByBookingId(bookingId);
        if (!intervals.isEmpty()) {
            intervalRepo.deleteAll(intervals);
            log.info("Cancelled calendar intervals for booking={} count={}", bookingId, intervals.size());
        }
    }

    // ── Register a booking directly (for iCal sync / manual reservations) ────

    @Transactional
    public CalendarInterval registerBookedInterval(UUID propertyId, LocalDate checkIn,
                                                    LocalDate checkOut, UUID bookingId, String note) {
        validateDates(checkIn, checkOut);
        acquireAdvisoryLock(propertyId);

        // Remove any stale RESERVED intervals for this exact range before inserting
        intervalRepo.findBlockingOverlap(propertyId, checkIn, checkOut).stream()
                .filter(ci -> "RESERVED".equals(ci.getState()))
                .forEach(intervalRepo::delete);

        List<CalendarInterval> remaining = intervalRepo.findBlockingOverlap(propertyId, checkIn, checkOut);
        if (!remaining.isEmpty()) {
            throw new AppException("Cannot register booking — dates already occupied", HttpStatus.CONFLICT);
        }

        CalendarInterval interval = CalendarInterval.builder()
                .propertyId(propertyId)
                .startDate(checkIn)
                .endDate(checkOut)
                .state("BOOKED")
                .bookingId(bookingId)
                .note(note)
                .build();
        interval = intervalRepo.save(interval);

        Property property = propertyRepo.findById(propertyId).orElse(null);
        if (property != null) {
            insertBuffers(interval);
        }

        return interval;
    }

    // ── Buffer engine ─────────────────────────────────────────────────────────

    private void insertBuffers(CalendarInterval booked) {
        Property property = propertyRepo.findById(booked.getPropertyId()).orElse(null);
        if (property == null) return;

        PropertyRuleResolver.ResolvedRules rules =
                ruleResolver.resolve(property, booked.getStartDate(), booked.getEndDate());

        if (rules.bufferDaysBefore() > 0) {
            LocalDate bufStart = booked.getStartDate().minusDays(rules.bufferDaysBefore());
            LocalDate bufEnd   = booked.getStartDate();
            // Only insert buffer if those dates are currently free
            if (intervalRepo.findBlockingOverlap(booked.getPropertyId(), bufStart, bufEnd).isEmpty()) {
                CalendarInterval buffer = CalendarInterval.builder()
                        .propertyId(booked.getPropertyId())
                        .startDate(bufStart)
                        .endDate(bufEnd)
                        .state("BUFFER")
                        .bookingId(booked.getBookingId())
                        .note("Auto-buffer before booking")
                        .build();
                intervalRepo.save(buffer);
            }
        }

        if (rules.bufferDaysAfter() > 0) {
            LocalDate bufStart = booked.getEndDate();
            LocalDate bufEnd   = booked.getEndDate().plusDays(rules.bufferDaysAfter());
            if (intervalRepo.findBlockingOverlap(booked.getPropertyId(), bufStart, bufEnd).isEmpty()) {
                CalendarInterval buffer = CalendarInterval.builder()
                        .propertyId(booked.getPropertyId())
                        .startDate(bufStart)
                        .endDate(bufEnd)
                        .state("BUFFER")
                        .bookingId(booked.getBookingId())
                        .note("Auto-buffer after booking")
                        .build();
                intervalRepo.save(buffer);
            }
        }
    }

    // ── Expiration cleanup (called by scheduler) ──────────────────────────────

    @Transactional
    public int expireStaleHolds() {
        List<BookingHold> expired = holdRepo.findExpired(Instant.now());
        int count = 0;
        for (BookingHold hold : expired) {
            if (hold.getIntervalId() != null) {
                intervalRepo.findById(hold.getIntervalId()).ifPresent(intervalRepo::delete);
            }
            hold.setStatus("EXPIRED");
            holdRepo.save(hold);
            count++;
        }
        if (count > 0) {
            log.info("Expired {} stale booking holds", count);
        }
        return count;
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private Property requireActiveProperty(UUID propertyId) {
        Property property = propertyRepo.findById(propertyId)
                .orElseThrow(() -> new AppException("Property not found", HttpStatus.NOT_FOUND));
        if (property.getStatus() != PropertyStatus.ACTIVE) {
            throw new AppException("Property is not accepting bookings (status: " + property.getStatus() + ")",
                    HttpStatus.BAD_REQUEST);
        }
        // Verify the organization has bookings enabled (passed host verification)
        boolean bookingsEnabled = hostVerificationRepo
                .findByOrganizationId(property.getOrganizationId())
                .map(HostVerification::isBookingsEnabled)
                .orElse(false);
        if (!bookingsEnabled) {
            throw new AppException(
                    "Bookings are disabled until verification is complete",
                    HttpStatus.FORBIDDEN);
        }
        return property;
    }

    private void validateDates(LocalDate start, LocalDate end) {
        if (start == null || end == null) {
            throw new AppException("Check-in and check-out dates are required", HttpStatus.BAD_REQUEST);
        }
        if (!end.isAfter(start)) {
            throw new AppException("Check-out must be after check-in", HttpStatus.BAD_REQUEST);
        }
        if (start.isBefore(LocalDate.now())) {
            throw new AppException("Check-in cannot be in the past", HttpStatus.BAD_REQUEST);
        }
    }

    private void validateStayRules(Property property, LocalDate checkIn, LocalDate checkOut) {
        PropertyRuleResolver.ResolvedRules rules = ruleResolver.resolve(property, checkIn, checkOut);
        long nights = java.time.temporal.ChronoUnit.DAYS.between(checkIn, checkOut);

        if (nights < rules.minStayDays()) {
            throw new AppException(
                    "Minimum stay is " + rules.minStayDays() + " night(s)", HttpStatus.BAD_REQUEST);
        }
        if (rules.maxStayDays() > 0 && nights > rules.maxStayDays()) {
            throw new AppException(
                    "Maximum stay is " + rules.maxStayDays() + " night(s)", HttpStatus.BAD_REQUEST);
        }
    }

    /**
     * Acquires a transaction-scoped advisory lock for the given property.
     * Uses hashtext() to convert UUID to an int8 key.
     * The lock is automatically released when the surrounding transaction commits or rolls back.
     * If the lock is held by another transaction, this call returns immediately with false
     * (non-blocking) and we throw a CONFLICT error to avoid deadlocks.
     */
    private void acquireAdvisoryLock(UUID propertyId) {
        String key = propertyId.toString();
        Boolean acquired = (Boolean) em.createNativeQuery(
                "SELECT pg_try_advisory_xact_lock(hashtext(:key))")
                .setParameter("key", key)
                .getSingleResult();

        if (Boolean.FALSE.equals(acquired)) {
            throw new AppException(
                    "Property calendar is busy — please retry in a moment", HttpStatus.CONFLICT);
        }
    }
}

File: frontend/src/api/calendarEngine.ts
import apiClient from './client'
import type { CalendarInterval, BookingHold, PropertyHouseRule, PropertySeasonalRule } from '@/types'
import { logger, shortId, maskEmail, maskName } from '@/lib/logger'

const log = logger.child('CALENDAR')

// ── Calendar Engine (interval-based availability) ─────────────────────────────

export const calendarEngineApi = {
  getCalendar: async (propertyId: string, from: string, to: string): Promise<CalendarInterval[]> => {
    log.debug('getCalendar — property=%s from=%s to=%s', shortId(propertyId), from, to)
    const { data } = await apiClient.get(`/properties/${propertyId}/calendar`, {
      params: { from, to }
    })
    log.debug('getCalendar — got %d intervals', data.data?.length)
    return data.data
  },

  blockDates: async (
    propertyId: string,
    startDate: string,
    endDate: string,
    reason?: string
  ): Promise<CalendarInterval> => {
    log.info('blockDates — property=%s %s→%s reason=%s', shortId(propertyId), startDate, endDate, reason)
    const { data } = await apiClient.post(`/properties/${propertyId}/calendar/block`, {
      startDate, endDate, reason
    })
    log.info('blockDates — success intervalId=%s', shortId(data.data?.id))
    return data.data
  },

  unblockDates: async (propertyId: string, intervalId: string): Promise<void> => {
    log.info('unblockDates — property=%s interval=%s', shortId(propertyId), shortId(intervalId))
    await apiClient.delete(`/properties/${propertyId}/calendar/intervals/${intervalId}`)
    log.info('unblockDates — success')
  },

  checkAvailability: async (
    propertyId: string,
    checkIn: string,
    checkOut: string
  ): Promise<{ available: boolean; reason?: string }> => {
    log.debug('checkAvailability — property=%s %s→%s', shortId(propertyId), checkIn, checkOut)
    const { data } = await apiClient.get(`/properties/${propertyId}/availability`, {
      params: { checkIn, checkOut }
    })
    log.info('checkAvailability — available=%s reason=%s', data.data?.available, data.data?.reason)
    return data.data
  },
}

// ── Public API (website builder / guest-facing) ───────────────────────────────

const publicBase = '/api/public'

export const publicCalendarApi = {
  getCalendar: async (propertyId: string, from: string, to: string): Promise<CalendarInterval[]> => {
    log.debug('public.getCalendar — property=%s from=%s to=%s', shortId(propertyId), from, to)
    const { data } = await apiClient.get(`${publicBase}/properties/${propertyId}/calendar`, {
      params: { from, to }
    })
    log.debug('public.getCalendar — got %d intervals', data.data?.length)
    return data.data
  },

  checkAvailability: async (
    propertyId: string,
    checkIn: string,
    checkOut: string
  ): Promise<{ available: boolean; reason?: string }> => {
    log.debug('public.checkAvailability — property=%s %s→%s', shortId(propertyId), checkIn, checkOut)
    const { data } = await apiClient.get(`${publicBase}/properties/${propertyId}/availability`, {
      params: { checkIn, checkOut }
    })
    log.info('public.checkAvailability — available=%s', data.data?.available)
    return data.data
  },

  createHold: async (
    propertyId: string,
    payload: {
      checkIn: string
      checkOut: string
      guests: number
      guestName?: string
      guestEmail?: string
      sessionId?: string
    }
  ): Promise<BookingHold> => {
    log.info('public.createHold — property=%s %s→%s guests=%d guestEmail=%s',
      shortId(propertyId),
      payload.checkIn,
      payload.checkOut,
      payload.guests,
      maskEmail(payload.guestEmail),
    )
    const { data } = await apiClient.post(`${publicBase}/properties/${propertyId}/holds`, payload)
    log.info('public.createHold — holdId=%s expiresAt=%s', shortId(data.data?.id), data.data?.expiresAt)
    return data.data
  },

  releaseHold: async (holdId: string): Promise<void> => {
    log.info('public.releaseHold — hold=%s', shortId(holdId))
    await apiClient.delete(`${publicBase}/holds/${holdId}`)
    log.info('public.releaseHold — success')
  },
}

// ── House rules ───────────────────────────────────────────────────────────────

export const houseRulesApi = {
  list: async (propertyId: string): Promise<PropertyHouseRule[]> => {
    log.debug('houseRules.list — property=%s', shortId(propertyId))
    const { data } = await apiClient.get(`/properties/${propertyId}/house-rules`)
    log.debug('houseRules.list — got %d rules', data.data?.length)
    return data.data
  },

  upsert: async (
    propertyId: string,
    ruleKey: string,
    allowed: boolean,
    notes?: string
  ): Promise<PropertyHouseRule> => {
    log.info('houseRules.upsert — property=%s rule=%s allowed=%s', shortId(propertyId), ruleKey, allowed)
    const { data } = await apiClient.put(`/properties/${propertyId}/house-rules`, {
      ruleKey, allowed, notes
    })
    return data.data
  },

  delete: async (propertyId: string, ruleId: string): Promise<void> => {
    log.info('houseRules.delete — property=%s rule=%s', shortId(propertyId), shortId(ruleId))
    await apiClient.delete(`/properties/${propertyId}/house-rules/${ruleId}`)
  },
}

// ── Seasonal rules ────────────────────────────────────────────────────────────

export const seasonalRulesApi = {
  list: async (propertyId: string): Promise<PropertySeasonalRule[]> => {
    log.debug('seasonalRules.list — property=%s', shortId(propertyId))
    const { data } = await apiClient.get(`/properties/${propertyId}/seasonal-rules`)
    log.debug('seasonalRules.list — got %d rules', data.data?.length)
    return data.data
  },

  create: async (
    propertyId: string,
    rule: Omit<PropertySeasonalRule, 'id' | 'propertyId'>
  ): Promise<PropertySeasonalRule> => {
    log.info('seasonalRules.create — property=%s name=%s %s→%s',
      shortId(propertyId), rule.name, rule.startDate, rule.endDate)
    const { data } = await apiClient.post(`/properties/${propertyId}/seasonal-rules`, rule)
    log.info('seasonalRules.create — success id=%s', shortId(data.data?.id))
    return data.data
  },

  delete: async (propertyId: string, ruleId: string): Promise<void> => {
    log.info('seasonalRules.delete — property=%s rule=%s', shortId(propertyId), shortId(ruleId))
    await apiClient.delete(`/properties/${propertyId}/seasonal-rules/${ruleId}`)
  },
}

// ── Amenities ─────────────────────────────────────────────────────────────────

export interface PropertyAmenity {
  id?: string
  propertyId?: string
  category: string
  name: string
  icon?: string
}

export const amenitiesApi = {
  list: async (propertyId: string): Promise<PropertyAmenity[]> => {
    log.debug('amenities.list — property=%s', shortId(propertyId))
    const { data } = await apiClient.get(`/properties/${propertyId}/amenities`)
    log.debug('amenities.list — got %d amenities', data.data?.length)
    return data.data
  },

  replace: async (propertyId: string, amenities: Omit<PropertyAmenity, 'id' | 'propertyId'>[]): Promise<PropertyAmenity[]> => {
    log.info('amenities.replace — property=%s count=%d', shortId(propertyId), amenities.length)
    const { data } = await apiClient.put(`/properties/${propertyId}/amenities`, amenities)
    log.info('amenities.replace — success')
    return data.data
  },

  delete: async (propertyId: string, amenityId: string): Promise<void> => {
    log.info('amenities.delete — property=%s amenity=%s', shortId(propertyId), shortId(amenityId))
    await apiClient.delete(`/properties/${propertyId}/amenities/${amenityId}`)
  },
}
